primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1010 @@
+		let public = self.0.to_public();
+		let mut raw = [0; PUBLIC_SERIALIZED_LEN];
+		public
+			.serialize_compressed(raw.as_mut_slice())
+			.expect("key buffer length is good; qed");
+		Public::unchecked_from(raw)
+	}
+
+	/// Sign raw data.
+	fn sign(&self, data: &[u8]) -> Signature {
+		let data = vrf::VrfSignData::new(SIGNING_CTX, &[data], vrf::VrfIosVec::default());
+		self.vrf_sign(&data).signature
+	}
+
+	/// Verify a signature on a message.
+	///
+	/// Returns `true` if the signature is good.
+	fn verify<M: AsRef<[u8]>>(signature: &Self::Signature, data: M, public: &Self::Public) -> bool {
+		let data = vrf::VrfSignData::new(SIGNING_CTX, &[data.as_ref()], vrf::VrfIosVec::default());
+		let signature =
+			vrf::VrfSignature { signature: *signature, vrf_outputs: vrf::VrfIosVec::default() };
+		public.vrf_verify(&data, &signature)
+	}
+
+	/// Return a vector filled with seed raw data.
+	fn to_raw_vec(&self) -> Vec<u8> {
+		// TODO @davxy: this function existance makes sense??? Should we return the seed or
+		// serialized secret key? If we return the serialized secret there is no method to
+		// reconstruct if ... unimplemented!()
+		panic!()
+	}
+}
+
+#[cfg(feature = "full_crypto")]
+impl CryptoType for Pair {
+	type Pair = Pair;
+}
+
+/// Bandersnatch VRF types and operations.
+pub mod vrf {
+	use super::*;
+	use crate::{bounded::BoundedVec, crypto::VrfCrypto, ConstU32};
+	use bandersnatch_vrfs::{
+		CanonicalDeserialize, CanonicalSerialize, IntoVrfInput, Message, PublicKey,
+		ThinVrfSignature, Transcript,
+	};
+
+	/// Max number of inputs/outputs which can be handled by the VRF signing procedures.
+	pub const MAX_VRF_IOS: u32 = 3;
+
+	/// Bounded vector used for VRF inputs and outputs.
+	///
+	/// Can contain at most [`MAX_VRF_IOS`] elements.
+	pub type VrfIosVec<T> = BoundedVec<T, ConstU32<MAX_VRF_IOS>>;
+
+	/// VRF input to construct a [`VrfOutput`] instance and embeddable within [`VrfSignData`].
+	#[derive(Clone, Debug)]
+	pub struct VrfInput(pub(super) bandersnatch_vrfs::VrfInput);
+
+	impl VrfInput {
+		/// Build a new VRF input.
+		///
+		/// Each message tuple has the form: message := (domain, data).
+		pub fn new(label: &'static [u8], messages: &[(&[u8], &[u8])]) -> Self {
+			// ⚠️ TODO @davxy @burdges (temporary hack and needs to be fixed)
+			// `bandersnatch_vrfs::Message` maps to a single (domain, data) tuple.
+			// We need something to push multiple messages together with a `label`.
+			// One solution is to construct the labeled Transcript here and then use
+			// `dleq_vrf::into_vrf_input()`.
+			// But has been commented out:
+			// https://github.com/w3f/ring-vrf/blob/8ab7b7b56e844f80b76afb1742b201fd69fb6046/dleq_vrf/src/vrf.rs#L38-L55
+			// THIS IS JUST A PLACEHOLDER and we are ignoring the lablel and concat the dom++msgs.
+			let _ = label;
+			let mut buf = Vec::new();
+			messages.into_iter().for_each(|(domain, message)| {
+				buf.extend_from_slice(domain);
+				buf.extend_from_slice(message);
```

Comment:
I think we need the code that has been commented out in `dleq_vrf`.
I.e. something to get a `VrfInput` from a **sequence** of messages.
I think that we should expose the `Transcript` structure in the backend so that we can construct it using the `label` passed by the user here.